proposals/stack-switching/Explainer.md

```diff
@@ -180,58 +186,51 @@ This abbreviation will be formalized with an auxiliary function or other means i
     - and `[t1*] -> [t2*] <: [t1'*] -> [t2'*]`
   - note - currently binding from right as discussed in https://github.com/WebAssembly/stack-switching/pull/53
 
-- `suspend <tagidx>`
-  - Send a tagged signal to suspend the current computation.
-  - `suspend $t : [t1*] -> [t2*]`
-    - iff `C.tags[$t] = tag $ft`
-    - and `C.types[$ft] ~~ func [t1*] -> [t2*]`
-
-- `resume <typeidx> (on <tagidx> <labelidx>|switch)*`
+- `resume <typeidx> hdl*`
   - Execute a given continuation.
     - If the executed continuation suspends with a tagged signal `$t`, the corresponding handler `(on $t H)` is executed.
-  - `resume $ct (on $t H)* : [t1* (ref null? $ct)] -> [t2*]`
+  - `resume $ct hdl* : [t1* (ref null? $ct)] -> [t2*]`
     - iff `C.types[$ct] ~~ cont [t1*] -> [t2*]`
-    - and for each `(tag $t H)`:
-      - `C.tags[$t] = tag $ft`
-      - and `C.types[$ft] ~~ func [te1*] -> [te2*]`
-      - and either `H = $l`
-        - and `C.labels[$l] = [te1'* (ref null? $ct')])*`
-        - and `([te1*] <: [te1'*])*`
-        - and `(C.types[$ct'] ~~ cont $ft')*`
-        - and `([te2*] -> [t2*] <: C.types[$ft'])*`
-      - or `H = switch`
-        - and `te1* = []`
-        - and `te2* <: t2*`
-
-- `resume_throw <typeidx> <exnidx> (on <tagidx> <labelidx>|switch)*`
+    - and `(hdl : t2*)*`
+
+- `resume_throw <typeidx> <exnidx> hdl*`
   - Execute a given continuation, but force it to immediately throw the annotated exception.
   - Used to abort a continuation.
-  - `resume_throw $ct $e (on $t H)* : [te* (ref null? $ct)] -> [t2*]`
+  - `resume_throw $ct $e hdl* : [te* (ref null? $ct)] -> [t2*]`
     - iff `C.types[$ct] ~~ cont [t1*] -> [t2*]`
-    - and `C.tags[$e] : tag $ft1`
-    - and `C.types[$ft1] ~~ func [te*] -> []`
-    - and for each `(tag $t H)`:
-      - `C.tags[$t] : tag $ft`
-      - and `C.types[$ft] ~~ func [te1*] -> [te2*]`
-      - and either `H = $l`
-        - and `C.labels[$l] = [te1'* (ref null? $ct')])*`
-        - and `([te1*] <: [te1'*])*`
-        - and `(C.types[$ct'] ~~ cont $ft')*`
-        - and `([te2*] -> [t2*] <: C.types[$ft'])*`
-      - or `H = switch`
-        - and `te1* = []`
-        - and `te2* <: t2*`
+    - and `C.tags[$e] : tag $ft`
+    - and `C.types[$ft] ~~ func [te*] -> []`
+    - and `(hdl : t2*)*`
+
+- `hdl = (on <tagidx> <labelidx>) | (on <tagidx> switch)`
+  - Handlers attached to `resume` and `resume_throw`, handling events for `suspend` and `switch`, respectively.
+  - `(on $e $l) : t*`
+    - iff `C.tags[$e] = tag $ft`
+    - and `C.types[$ft] ~~ func [t1*] -> [t2*]`
+    - and `C.labels[$l] = [t1'* (ref null? $ct)]`
+    - and `t1* <: t1'*`
+    - and `C.types[$ct] ~~ cont [t2'*] -> [t'*]`
+    - and `[t2*] -> [t*] <: [t2'*] -> [t'*]`
```

Comment:
Actually, I think this subtype must go the other way: the used continuation type must respect, i.e. be a subtype of, the types required by the handler.

Reply:
Right you are, thanks.

Reply:
Wait, `t2*` is the type the continuation is actually expecting to receive and `t2'*` is the type we are assigning the continuation parameters. If `t2'*` is a strict supertype of `t2*`, then we would be able to send the continuation a value it's not expecting. So `[t2'*] -> [t'*] <: [t2*] -> [t*]` can't be right either.

If `t2'*` is a strict subtype of of `t2*`, then we can still only send the continuation values it's expecting, so that can be allowed.

Similarly, `t*` is the type the continuation is actually returning and `t'*` is the type we are assigning its return values. So everything is fine when `t* <: t'*`.

So the correct constraints here are

```
t2'* <: t2*
t* <: t'*`
```

In other words, `[t2*] -> [t*] <: [t2'*] -> [t'*]`, the current constraint, is correct.

Would it be clearer if we split it up into the two separate constraints on resulttypes instead of having a single constraint on a functype?

Reply:
Ah, I think you're right, I managed to confuse myself about the direction of the dataflow. Just keep it as is.